src/extension.ts

```diff
@@ -83,12 +84,30 @@ export function activate(context: ExtensionContext) {
 
     executeAmebaOnWorkspace(ameba);
 
+    // This can happen when a file is open _or_ when a file's language id changes
     workspace.onDidOpenTextDocument(doc => {
-        ameba && ameba.execute(doc);
+        if (ameba && ameba.config.trigger !== LintTrigger.None && isCrystalDocument(doc)) {
+            if (isDocumentVirtual(doc)) {
+                if (ameba.config.trigger === LintTrigger.Type) {
+                    outputChannel.appendLine(`[Open] Running ameba on ${getRelativePath(doc)}`);
+                    ameba.execute(doc, true);
+                }
+            } else {
+                outputChannel.appendLine(`[Open] Running ameba on ${getRelativePath(doc)}`);
+                ameba.execute(doc);
+            }
+        }
     });
 
+    workspace.onDidChangeTextDocument(e => {
+        if (ameba && ameba.config.trigger == LintTrigger.Type && isCrystalDocument(e.document)) {
+            outputChannel.appendLine(`[Change] Running ameba on ${getRelativePath(e.document)}`);
+            ameba.execute(e.document, true);
+        }
+    })
+
     workspace.onDidSaveTextDocument(doc => {
-        if (ameba && ameba.config.onSave && isValidCrystalDocument(doc)) {
+        if (ameba && ameba.config.trigger === LintTrigger.Save && isCrystalDocument(doc)) {
```

Comment:
AFAIU this should fire for both triggers:

```suggestion
        if (ameba && ameba.config.trigger !== LintTrigger.None && isCrystalDocument(doc)) {
```

Reply:
onDidChangeTextDocument runs when a document is saved, so this callback should only handle the case where we're only linting when saving.

Reply:
You're right, had a brain fog while commenting on this one... 🤦‍♂️